### PR TITLE
Upgrading to Spring Boot 3

### PR DESCRIPTION
**Pull request by kinow**
## Description

Upgrading to Spring Boot 3

## Motivation and Context

https://github.com/common-workflow-language/cwlviewer/pull/566#issuecomment-1735611928

Closes #576 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-version: '3.9'
+version: '3.2'
```
Comment (kinow) [inline docker-compose.yml:1]: To be reverted

File: src/main/java/org/commonwl/view/cwl/CWLValidationException.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 package org.commonwl.view.cwl;
 
-import javax.validation.ValidationException;
+import jakarta.validation.ValidationException;
```
Comment (kinow) [inline src/main/java/org/commonwl/view/cwl/CWLValidationException.java:22]: Note: Iacopo might have to test a deploy of this version, to make sure the server is using jakarta and not javax too.

File: src/main/java/org/commonwl/view/util/BaseEntity.java [hunk 3]
```diff
@@ -1,10 +1,6 @@
 package org.commonwl.view.util;
 
-import com.vladmihalcea.hibernate.type.json.JsonBinaryType;
-import javax.persistence.MappedSuperclass;
-import org.hibernate.annotations.TypeDef;
-import org.hibernate.annotations.TypeDefs;
+import jakarta.persistence.MappedSuperclass;
 
-@TypeDefs({@TypeDef(name = "json", typeClass = JsonBinaryType.class)})
```
Comment (kinow) [inline src/main/java/org/commonwl/view/util/BaseEntity.java:8]: Not needed with Hibernate 6.

File: src/main/java/org/commonwl/view/workflow/QueuedWorkflow.java [hunk 4]
```diff
@@ -2,16 +2,17 @@
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 import com.fasterxml.jackson.annotation.JsonInclude;
+import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
```
Comment (kinow) [inline src/main/java/org/commonwl/view/workflow/QueuedWorkflow.java:5]: They renamed the package hibernate-entity-60, to this hypersistence.

File: src/main/java/org/commonwl/view/workflow/WorkflowController.java [hunk 5]
```diff
@@ -214,7 +214,7 @@ public ModelAndView getWorkflow(
    * @param branch The branch of the repository
    * @return The workflow view with the workflow as a model
    */
-  @GetMapping(value = "/workflows/**/*.git/{branch}/**")
+  @GetMapping(value = "/workflows/*/*.git/{branch}/**")
```
Comment (kinow) [inline src/main/java/org/commonwl/view/workflow/WorkflowController.java:217]: This needs to be well tested.

Now Spring Web doesn't allow any content after the `/**`. I found an issue or blog that suggested to test with `/*/bla` instead. Trying to get the tests and build to pass in order to verify it in the browser.

File: src/main/resources/logback.xml [hunk 6]
```diff
@@ -9,7 +9,7 @@
             class="ch.qos.logback.core.ConsoleAppender">
     <layout class="ch.qos.logback.classic.PatternLayout">
       <Pattern>
-        %black(%d{ISO8601}) %highlight(%-5level) [%blue(%t)] %yellow(%C{1.}): %msg%n%throwable
+        %black(%d{ISO8601}) %highlight(%-5level) [%blue(%t)] %yellow(%C{1}): %msg%n%throwable
```
Comment (kinow) [inline src/main/resources/logback.xml:12]: Was getting a `NumberFormatException` for `1.`, during the logback startup. This was the only place where I found the `1.`. Changing it the build didn't fail in the same place, but will have to confirm what was the 1., and if the 1 alone will act the same way.

**Comment by kinow** [comment]
Fixed one broken test today, three more to go. Two I know the reason, but not the solution. The other we had during the migration mongo->postgres, so I just need to find the notes about how to debug that one (need to look at the query executed during the unit test...)

**Comment by kinow** [comment]
The last test failing is due to Hibernate/Postgres. At least the others were fixed (at least locally, waiting for GH Actions). Time to search for issues/SO's/etc now :grimacing:

File: src/main/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryImpl.java [hunk 7]
```diff
@@ -19,12 +19,19 @@ public class QueuedWorkflowRepositoryImpl implements QueuedWorkflowRepositoryCus
...
   @Override
   public QueuedWorkflow findByRetrievedFrom(GitDetails retrievedFrom) {
+    // N.B. The migration from Spring Boot 2 to 3 got blocked for a while as
+    //      we couldn't figure out a way to get the mapping of types to work.
+    //      We always ended up the error about the invalid operator for the
+    //      jsonb = bytea types. Found this comment from the author of the
+    //      mapping library we use, with what was the fix for our issue (ran
+    //      out of ideas, so started testing everything found online):
+    //      Use `new JsonType(GitDetails.class)`. That finally solved it.
+    //      Ref: https://github.com/common-workflow-language/cwlviewer/pull/568
```
Comment (kinow) [inline src/main/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryImpl.java:29]: The fix, finally :point_up: 

Works on my computer™, but pending tests with some workflows to confirm it's working.

**Comment by kinow** [comment]
Finally had some time to give it another try, and managed to get all tests passing. Still a couple of points pending, and I will try to do some manual tests using existing CWL workflows. But feel free to review or test it if you have time -- I will test it next time I have a chance, but at least we now have a working build with updated Spring Boot and dependencies :wave:

File: src/test/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryTest.java [hunk 8]
```diff
@@ -5,47 +5,35 @@
...
-@Transactional(propagation = Propagation.REQUIRED)
+@ContextConfiguration(initializers = PostgreSQLContextInitializer.class)
+// @EnableJpaRepositories("org.commonwl.view.workflow")
+// @EntityScan
+// @Transactional(propagation = Propagation.REQUIRED)
+// @ContextConfiguration(
+//    initializers = PostgreSQLContextInitializer.class,
+//    classes = {WebConfig.class, CwlViewerApplication.class, QueuedWorkflowRepository.class})
+// @SpringBootTest(classes = CwlViewerApplication.class)
+// @RunWith(SpringRunner.class)
```
Comment (mr-c) [inline src/test/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryTest.java:30]: What's the story here?

**Review by mr-c — COMMENTED** [review]
Really exciting, thank you @kinow !

File: src/test/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryTest.java (hunk 8 above)
Comment (kinow) [reply src/test/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryTest.java:30]: I used this to test, but I can remove this. I don't think I, or any other developer, will need this. Removing it...

Comment (kinow) [reply src/test/java/org/commonwl/view/workflow/QueuedWorkflowRepositoryTest.java]: Sorry, got distracted by something. Done, and all tests passed.

**Comment by mr-c** [comment]
@kinow Shall I merge?

**Comment by kinow** [comment]
> @kinow Shall I merge?

Let me rebase, and then do some quick tests with Docker locally. Then it should be ready to go! :boom:

**Comment by kinow** [comment]
@mr-c rebased, GH Actions passed, then checked out latest commit, ran `docker compose up`, uploaded some CWL files, they were all correctly rendered. I think it's ready to be merged now :slightly_smiling_face: Thanks!

**Comment by mr-c** [comment]
Huzzah!

**Pull request merged**